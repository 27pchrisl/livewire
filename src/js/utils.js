
export function dispatch(eventName, { target, cancelable, data } = {}) {
    const event = document.createEvent("Events")
    event.initEvent(eventName, true, cancelable == true)
    event.data = data || {}

    // Fix setting `defaultPrevented` when `preventDefault()` is called
    // http://stackoverflow.com/questions/23349191/event-preventdefault-is-not-working-in-ie-11-for-custom-events
    if (event.cancelable && !preventDefaultSupported) {
        const { preventDefault } = event
        event.preventDefault = function () {
            if (!this.defaultPrevented) {
                Object.defineProperty(this, "defaultPrevented", { get: () => true })
            }
            preventDefault.call(this)
        }
    }

    (target || document).dispatchEvent(event)
    return event
}

const preventDefaultSupported = (() => {
    const event = document.createEvent("Events")
    event.initEvent("test", true, true)
    event.preventDefault()
    return event.defaultPrevented
})()

export function addMixin(classTarget, ...sources) {
    sources.forEach(source => {
        let descriptors = Object.keys(source).reduce((descriptors, key) => {
            descriptors[key] = Object.getOwnPropertyDescriptor(source, key);
            return descriptors;
        }, {});

        Object.getOwnPropertySymbols(source).forEach(sym => {
            let descriptor = Object.getOwnPropertyDescriptor(source, sym);
            if (descriptor.enumerable) {
                descriptors[sym] = descriptor;
            }
        });
        Object.defineProperties(classTarget.prototype, descriptors);
    });
    return classTarget.prototype;
}
