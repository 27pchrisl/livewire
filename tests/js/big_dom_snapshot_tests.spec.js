import { fireEventAndMakeServerRespondWithDom } from './utils'

test('test basic click', () => {
    document.body.innerHTML = `<div id="app">
        <div wire:id="componentA" wire:serialized="empty">
            <form wire:submit="doSomething" wire:ref="submitEl">
                <div id="spinner" class="hidden" wire:loading="submitEl"></div>
                <button></button>
            </form>
        </div>
    </div>`

    fireEventAndMakeServerRespondWithDom('button', 'click', document.querySelector(`[wire\\:id="componentA"]`).outerHTML)

    expect(document.body.outerHTML).toMatchSnapshot();
})

test('test adding element', () => {
    document.body.innerHTML = `<div id="app">
        <div wire:id="componentA" wire:serialized="empty">
            <form wire:submit="doSomething" wire:ref="submitEl">
                <div id="spinner" class="hidden" wire:loading="submitEl"></div>
                <button></button>
            </form>
        </div>
    </div>`

    fireEventAndMakeServerRespondWithDom('button', 'click',
        `<div wire:id="componentA" wire:serialized="empty">
            <form wire:submit="doSomething" wire:ref="submitEl">
                <div id="spinner" class="hidden" wire:loading="submitEl"></div>
                <span>Something updated</span>
                <button>hey</button>
            </form>
        </div>`
    )

    expect(document.body.outerHTML).toMatchSnapshot();
})

test('test adding element with sibling components', () => {
    document.body.innerHTML = `<div id="app">
        <div wire:id="componentA" wire:serialized="empty">
            <form wire:submit="doSomething" wire:ref="submitEl">
                <div id="spinner" class="hidden" wire:loading="submitEl"></div>
                <button></button>
            </form>
        </div>

        <div wire:id="componentB" wire:serialized="empty">
            <form wire:submit="doSomething" wire:ref="submitEl">
                <div id="spinner" class="hidden" wire:loading="submitEl"></div>
                <button></button>
            </form>
        </div>
    </div>`

    fireEventAndMakeServerRespondWithDom('[wire:id="componentB"] button', 'click',
        `<div wire:id="componentB" wire:serialized="empty">
            <form wire:submit="doSomething" wire:ref="submitEl">
                <div id="spinner" class="hidden" wire:loading="submitEl"></div>
                <button>button</button>
                <div>something added</div>
            </form>
        </div>`
    )

    expect(document.body.outerHTML).toMatchSnapshot();
})

test('test adding element inside nested component', () => {
    document.body.innerHTML = `<div id="app">
        <div wire:id="componentA" wire:serialized="empty">
            <form wire:submit="doSomething" wire:ref="submitEl">
                <div id="spinner" class="hidden" wire:loading="submitEl"></div>
                <button></button>
            </form>

            <div wire:id="componentB" wire:serialized="empty">
                <form wire:submit="doSomething" wire:ref="submitEl">
                    <div id="spinner" class="hidden" wire:loading="submitEl"></div>
                    <button></button>
                </form>
            </div>
        </div>
    </div>`

    fireEventAndMakeServerRespondWithDom('[wire:id="componentB"] button', 'click',
        `<div wire:id="componentB" wire:serialized="empty">
            <form wire:submit="doSomething" wire:ref="submitEl">
                <div id="spinner" class="hidden" wire:loading="submitEl"></div>
                <button>button</button>
                <div>something added</div>
            </form>
        </div>`
    )

    expect(document.body.outerHTML).toMatchSnapshot();
})

test('test adding element outside nested component', () => {
    document.body.innerHTML = `<div id="app">
        <div wire:id="componentA" wire:serialized="empty">
            <form wire:submit="doSomething" wire:ref="submitEl">
                <div id="spinner" class="hidden" wire:loading="submitEl"></div>
                <button></button>
            </form>

            <div wire:id="componentB" wire:serialized="empty">
                <form wire:submit="doSomething" wire:ref="submitEl">
                    <div id="spinner" class="hidden" wire:loading="submitEl"></div>
                    <button></button>
                </form>
            </div>
        </div>
    </div>`

    fireEventAndMakeServerRespondWithDom('[wire:id="componentA"] button', 'click',
        `<div wire:id="componentA" wire:serialized="empty">
            <form wire:submit="doSomething" wire:ref="submitEl">
                <div id="spinner" class="hidden" wire:loading="submitEl"></div>
                <button>button</button>
                <div>something added</div>
            </form>

            <div wire:id="componentB" wire:serialized="empty">
            </div>
        </div>`
    )

    expect(document.body.outerHTML).toMatchSnapshot();
})

test('test adding nested component', () => {
    document.body.innerHTML = `<div id="app">
        <div wire:id="componentA" wire:serialized="empty">
            <form wire:submit="doSomething" wire:ref="submitEl">
                <div id="spinner" class="hidden" wire:loading="submitEl"></div>
                <button>button</button>
            </form>
        </div>
    </div>`

    fireEventAndMakeServerRespondWithDom('[wire:id="componentA"] button', 'click',
        `<div wire:id="componentA" wire:serialized="empty">
            <form id="yih" wire:submit="doSomething" wire:ref="submitEl">
                <div id="spinner" class="hidden" wire:loading="submitEl"></div>
                <button>button</button>

                <div wire:id="componentB" wire:serialized="empty">
                    <div id="yo" wire:submit="doSomething" wire:ref="submitEl">
                        <div id="spinner" class="hidden" wire:loading="submitEl"></div>
                    </div>
                </div>
            </form>
        </div>`
    )

    expect(document.body.outerHTML).toMatchSnapshot();
})

test('test removing nested component', () => {
    document.body.innerHTML = `<div id="app">
        <div wire:id="componentA" wire:serialized="empty">
            <form wire:submit="doSomething" wire:ref="submitEl">
                <div id="spinner" class="hidden" wire:loading="submitEl"></div>
                <button></button>
            </form>

            <div wire:id="componentB" wire:serialized="empty">
                <form wire:submit="doSomething" wire:ref="submitEl">
                    <div id="spinner" class="hidden" wire:loading="submitEl"></div>
                    <button></button>
                </form>
            </div>
        </div>
    </div>`

    fireEventAndMakeServerRespondWithDom('[wire:id="componentA"] button', 'click',
        `<div wire:id="componentA" wire:serialized="empty">
            <form wire:submit="doSomething" wire:ref="submitEl">
                <div id="spinner" class="hidden" wire:loading="submitEl"></div>
                <button>button</button>
                <div>something added</div>
            </form>
        </div>`
    )

    expect(document.body.outerHTML).toMatchSnapshot();
})
