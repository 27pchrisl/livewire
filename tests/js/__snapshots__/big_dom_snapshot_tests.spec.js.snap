// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test adding element 1`] = `
"<body><div id=\\"app\\">
        <div wire:root-id=\\"componentA\\" wire:root-serialized=\\"empty\\">
            <form wire:submit=\\"doSomething\\" wire:ref=\\"submitEl\\">
                <div id=\\"spinner\\" class=\\"hidden\\" wire:loading=\\"submitEl\\"></div>
                <span>Something updated</span>
                <button>hey</button>
            </form>
        </div>
    </div></body>"
`;

exports[`test adding element inside nested component 1`] = `
"<body><div id=\\"app\\">
        <div wire:root-id=\\"componentA\\" wire:root-serialized=\\"empty\\">
            <form wire:submit=\\"doSomething\\" wire:ref=\\"submitEl\\">
                <div id=\\"spinner\\" class=\\"hidden\\" wire:loading=\\"submitEl\\"></div>
                <button></button>
            </form>

            <div wire:root-id=\\"componentB\\" wire:root-serialized=\\"empty\\">
            <form wire:submit=\\"doSomething\\" wire:ref=\\"submitEl\\">
                <div id=\\"spinner\\" class=\\"hidden\\" wire:loading=\\"submitEl\\"></div>
                <button>button</button>
                <div>something added</div>
            </form>
        </div>
        </div>
    </div></body>"
`;

exports[`test adding element outside nested component 1`] = `
"<body><div id=\\"app\\">
        <div wire:root-id=\\"componentA\\" wire:root-serialized=\\"empty\\">
            <form wire:submit=\\"doSomething\\" wire:ref=\\"submitEl\\">
                <div id=\\"spinner\\" class=\\"hidden\\" wire:loading=\\"submitEl\\"></div>
                <button>button</button>
                <div>something added</div>
            </form>

            <div wire:root-id=\\"componentB\\" wire:root-serialized=\\"empty\\">
                <form wire:submit=\\"doSomething\\" wire:ref=\\"submitEl\\">
                    <div id=\\"spinner\\" class=\\"\\" wire:loading=\\"submitEl\\"></div>
                    <button></button>
                </form>
            </div>
        </div>
    </div></body>"
`;

exports[`test adding element with sibling components 1`] = `
"<body><div id=\\"app\\">
        <div wire:root-id=\\"componentA\\" wire:root-serialized=\\"empty\\">
            <form wire:submit=\\"doSomething\\" wire:ref=\\"submitEl\\">
                <div id=\\"spinner\\" class=\\"hidden\\" wire:loading=\\"submitEl\\"></div>
                <button></button>
            </form>
        </div>

        <div wire:root-id=\\"componentB\\" wire:root-serialized=\\"empty\\">
            <form wire:submit=\\"doSomething\\" wire:ref=\\"submitEl\\">
                <div id=\\"spinner\\" class=\\"hidden\\" wire:loading=\\"submitEl\\"></div>
                <button>button</button>
                <div>something added</div>
            </form>
        </div>
    </div></body>"
`;

exports[`test basic click 1`] = `
"<body><div id=\\"app\\">
        <div wire:root-id=\\"componentA\\" wire:root-serialized=\\"empty\\">
            <form wire:submit=\\"doSomething\\" wire:ref=\\"submitEl\\">
                <div id=\\"spinner\\" class=\\"\\" wire:loading=\\"submitEl\\"></div>
                <button></button>
            </form>
        </div>
    </div></body>"
`;

exports[`test removing nested component 1`] = `
"<body><div id=\\"app\\">
        <div wire:root-id=\\"componentA\\" wire:root-serialized=\\"empty\\">
            <form wire:submit=\\"doSomething\\" wire:ref=\\"submitEl\\">
                <div id=\\"spinner\\" class=\\"hidden\\" wire:loading=\\"submitEl\\"></div>
                <button>button</button>
                <div>something added</div>
            </form>
        </div>
    </div></body>"
`;
